File: files/CH-096.R
library(tidyverse)
library(readxl)

path = "files/CH-96 Top Products.xlsx"
input = read_excel(path, range = "B2:D20")
test  = read_excel(path, range = "I2:K11")

result = input %>%
  mutate(Month = month(Date)) %>%
  summarise(Quantity = sum(Quantity), .by = c(Product, Month)) %>%
  mutate(Rank = dense_rank(desc(Quantity)), .by = Month) %>%
  mutate(Product = ifelse(Rank < 3, Product, "Other")) %>%
  arrange(Month, Rank) %>%
  summarise(Quantity = sum(Quantity), .by = c(Product, Month)) %>%
  mutate(`% of Month sales` = Quantity / sum(Quantity), .by = Month) %>%
  select(Month, Product, `% of Month sales`)

identical(result, test)
# [1] TRUE

File: files/000-099/096/CH-096.py
import pandas as pd

path = "CH-96 Top Products.xlsx"
input = pd.read_excel(path, skiprows=1, usecols="B:D", nrows=19)
test = pd.read_excel(path, skiprows=1, usecols="I:K", nrows=9)\
    .rename(columns={"Product.1":"Product", "% of Month sales":"Percent"})\
    .sort_values(["Month","Product"])\
    .reset_index(drop=True)

result = input.assign(Month=input['Date'].dt.month) \
    .groupby(['Product', 'Month']) \
    .agg(Quantity=('Quantity', 'sum')) \
    .reset_index() \
    .assign(Rank=lambda x: x.groupby('Month')['Quantity'].rank(method='dense', ascending=False)) \
    .assign(Product=lambda x: x.apply(lambda row: row['Product'] if row['Rank'] < 3 else 'Other', axis=1)) \
    .sort_values(['Month', 'Rank']) \
    .groupby(['Product', 'Month']) \
    .agg(Quantity=('Quantity', 'sum')) \
    .sort_values(["Month","Product"])\
    .reset_index() \
    .assign(Percent=lambda x: x['Quantity'] / x.groupby('Month')['Quantity'].transform('sum')) \
    .loc[:, ['Month', 'Product', 'Percent']]

print(all(result == test)) # True